server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1052,7 +1053,13 @@ protected void doRun() {
             public void onFailure(final Exception e) {
                 Snapshot snapshot = entry.snapshot();
                 logger.warn(() -> new ParameterizedMessage("[{}] failed to finalize snapshot", snapshot), e);
-                removeSnapshotFromClusterState(snapshot, null, e);
+                if (ExceptionsHelper.unwrap(e, NotMasterException.class) != null) {
```

Comment:
This feels kind of dirty but IMO it's an ok best-effort work around for now. We simply aren't exposing the internals of the cluster state updates from the `BlobStoreRepository` and I don't see a straight-forward way of doing so with the current `Repository` interface. Also, this is just to help (as in improve the user experience, there's no corruption to be fixed here :)) with the incredibly unlikely corner case showing in the linked test failure were a new master is elected, immediately fails over again and is then finally elected for good.

Reply:
You should also handle `FailedToCommitClusterStateException` here (see also TransportMasterNodeAction). Note that I'm not sure how the wrapping comes into play here / whether we need to unwrap.

Reply:
> Note that I'm not sure how the wrapping comes into play here / whether we need to unwrap.

Not sure either, so unwrapping seems safer to me

Reply:
Right, added that exception here :)

We gotta unwrap, we're wrapping everything in the CS update callbacks in a `RepositoryException` in `BlobStoreRepository`

Reply:
`RepositoryException` is a not a `WrapperException`, so the unwrapping won't work. I'm confused how this fixes the test.

Reply:
I might be missing something, but this fix uses `unwrap()` which acts differently from `unwrapCause()` and does not rely on ElasticsearchWrapperException.

Reply:
`org.elasticsearch.ExceptionsHelper#unwrap` just loops through the `getCause` returns, it doesn't need `WrapperException` like `org.elasticsearch.ExceptionsHelper#unwrapCause`?

Reply:
lol, thanks for the clarification. What a confusing bunch of methods :D